ortools/math_opt/solvers/xpress_solver.cc

```diff
@@ -0,0 +1,693 @@
+      model_parameters, kXpressSupportedStructures, "XPRESS"));
+  const absl::Time start = absl::Now();
+
+  RETURN_IF_ERROR(CheckRegisteredCallbackEvents(callback_registration,
+                                                /*supported_events=*/{}));
+
+  RETURN_IF_ERROR(CheckParameters(parameters));
+
+  // Check that bounds are not inverted just before solve
+  // XPRESS returns "infeasible" when bounds are inverted
+  {
+    ASSIGN_OR_RETURN(const InvertedBounds inv_bounds, ListInvertedBounds());
+    RETURN_IF_ERROR(inv_bounds.ToStatus());
+  }
+
+  // Set initial basis
+  if (model_parameters.has_initial_basis()) {
+    RETURN_IF_ERROR(SetXpressStartingBasis(model_parameters.initial_basis()));
+  }
+
+  RETURN_IF_ERROR(CallXpressSolve(parameters)) << "Error during XPRESS solve";
+
+  ASSIGN_OR_RETURN(SolveResultProto solve_result,
+                   ExtractSolveResultProto(start, model_parameters));
+
+  return solve_result;
+}
+
+inline std::string GetOptimizationFlags(
+    const SolveParametersProto& parameters) {
+  switch (parameters.lp_algorithm()) {
+    case LP_ALGORITHM_PRIMAL_SIMPLEX:
+      return "p";
+    case LP_ALGORITHM_DUAL_SIMPLEX:
+      return "d";
+    case LP_ALGORITHM_BARRIER:
+      return "b";
+    default:
+      // this makes XPRESS use default algorithm (XPRS_DEFAULTALG)
+      return "";
+  }
+}
+absl::Status XpressSolver::CallXpressSolve(
+    const SolveParametersProto& parameters) {
+  // Enable screen output right before solve
+  if (parameters.enable_output()) {
+    RETURN_IF_ERROR(xpress_->SetIntAttr(XPRS_OUTPUTLOG, 1))
+        << "Unable to enable XPRESS logs";
+  }
+  // Solve
+  if (is_mip_) {
+    ASSIGN_OR_RETURN(xpress_status_, xpress_->MipOptimizeAndGetStatus());
+  } else {
+    RETURN_IF_ERROR(SetLpIterLimits(parameters))
+        << "Could not set iteration limits.";
+    ASSIGN_OR_RETURN(xpress_status_, xpress_->LpOptimizeAndGetStatus(
+                                         GetOptimizationFlags(parameters)));
+  }
+  // Post-solve
+  if (!(is_mip_ ? (xpress_status_ == XPRS_MIP_OPTIMAL)
+                : (xpress_status_ == XPRS_LP_OPTIMAL))) {
+    RETURN_IF_ERROR(xpress_->PostSolve()) << "Post-solve failed in XPRESS";
+  }
+  // Disable screen output right after solve
+  if (parameters.enable_output()) {
+    RETURN_IF_ERROR(xpress_->SetIntAttr(XPRS_OUTPUTLOG, 0))
+        << "Unable to disable XPRESS logs";
+  }
+  return absl::OkStatus();
+}
+
+absl::Status XpressSolver::SetLpIterLimits(
+    const SolveParametersProto& parameters) {
+  // If the user has set no limits, we still have to reset the limits
+  // explicitly to their default values, else the parameters could be kept
+  // between solves. XPRESS default value for XPRS_LPITERLIMIT is 2147483647
+  int lp_iter_limit = parameters.has_iteration_limit()
+                          ? parameters.iteration_limit()
+                          : 2147483647;
+  // XPRESS default value for XPRS_BARITERLIMIT is 500
```

Comment:
In order to get the default values, you should rather query the attributes right after calling `XPRScreateprob()`. This seems more future proof.